symengine/fields.cpp

```diff
@@ -462,4 +462,112 @@ GaloisFieldDict::gf_ddf_zassenhaus() const
     }
     return factors;
 }
+
+GaloisFieldDict
+GaloisFieldDict::_gf_pow_pnm1d2(const GaloisFieldDict &f,
+                                const integer_class &n,
+                                const std::vector<GaloisFieldDict> &b) const
+{
+    GaloisFieldDict f_in(f);
+    f_in %= *this;
+    GaloisFieldDict h, r;
+    h = r = f_in;
+    for (unsigned i = 1; i < n; ++i) {
+        h = h.gf_frobenius_map(*this, b);
+        r *= h;
+        r %= *this;
+    }
+    auto res = gf_pow_mod(r, (modulo_ - 1_z) / 2_z);
+    return res;
+}
+
+GaloisFieldDict GaloisFieldDict::gf_random(const unsigned long &n_val,
+                                           gmp_randstate_t &state) const
+{
+    std::vector<integer_class> v(n_val + 1);
+    for (int i = 0; i < n_val; ++i) {
+        integer_class temp;
+        mp_urandomm(temp, state, modulo_);
+        v[i] = temp;
+    }
+    v[n_val] = 1_z;
+    return GaloisFieldDict::from_vec(v, modulo_);
+}
+
+std::set<GaloisFieldDict, GaloisFieldDict::DictLess>
+GaloisFieldDict::gf_edf_zassenhaus(const integer_class &n) const
+{
+    std::set<GaloisFieldDict, DictLess> factors;
+    factors.insert(*this);
+    unsigned n_val = mp_get_si(n);
+    if (this->degree() <= n_val)
+        return factors;
+
+    unsigned N = this->degree() / n_val;
+
+    std::vector<GaloisFieldDict> b;
+    if (modulo_ != 2_z)
+        b = this->gf_frobenius_monomial_base();
+    gmp_randstate_t state;
+    gmp_randinit_default(state);
+    gmp_randseed_ui(state, std::rand());
+    while (factors.size() < N) {
+        auto r = gf_random(2 * n_val - 1, state);
+        GaloisFieldDict g;
+        if (modulo_ == 2_z) {
```

Comment:
Well actually no, I think we are getting lucky in case of `gf_edf_zassenhaus([3,0,1],1,4,ZZ)`, because we take gcd with a random number and its leading coefficient might not be invertible in the `modulo_